crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move

```diff
@@ -0,0 +1,1045 @@
+    /// - Emits a `StartDKGFirstRoundEvent`.
+    ///
+    /// ### Emits
+    /// - `StartDKGFirstRoundEvent`:
+    ///   - `session_id`: The generated session ID.
+    ///   - `initiator`: The address of the transaction sender.
+    ///   - `dwallet_cap_id`: The ID of the created `DWalletCap`.
+    public fun launch_dkg_first_round(ctx: &mut TxContext) {
+        let dwallet_cap = create_dwallet_cap(ctx);
+        let dwallet_cap_id = object::id(&dwallet_cap);
+        transfer::public_transfer(dwallet_cap, tx_context::sender(ctx));
+        let initiator = tx_context::sender(ctx);
+        let session_id = tx_context::fresh_object_address(ctx);
+        event::emit(StartDKGFirstRoundEvent {
+            session_id,
+            initiator,
+            dwallet_cap_id,
+        });
+    }
+
+    /// Creates the output of the first DKG round.
+    ///
+    /// This function transfers the output of the first DKG round
+    /// to the session initiator and ensures it is securely linked
+    /// to the `DWalletCap` of the session.
+    /// This function is called by blockchain itself.
+    /// Validators call it, it's part of the blockchain logic.
+    ///
+    /// ### Effects
+    /// - Transfers the output of the first round to the initiator.
+    /// - Emits necessary metadata and links it to the associated session.
+    ///
+    /// ### Parameters
+    /// - `initiator`: The address of the user who initiated the DKG session.
+    /// - `session_id`: The ID of the DKG session.
+    /// - `output`: The output data from the first round.
+    /// - `dwallet_cap_id`: The ID of the associated `DWalletCap`.
+    /// - `ctx`: The transaction context.
+    ///
+    /// ### Panics
+    /// - Panics with `ENotSystemAddress` if the sender is not the system address.
+    #[allow(unused_function)]
+    /// Creates the output of the first round of the DKG MPC, transferring it to the initiating user.
+    /// This function is called by the blockchain itself.
+    /// Validators call it as part of the blockchain logic.
+    fun create_dkg_first_round_output(
+        session_id: ID,
+        output: vector<u8>,
+        ctx: &TxContext
+    ) {
+        assert!(tx_context::sender(ctx) == SYSTEM_ADDRESS, ENotSystemAddress);
+        event::emit(DKGFirstRoundOutputEvent {
+            session_id,
+            output,
+        });
+    }
+
+    /// Starts the second DKG round.
+    /// Emits an event for validators to begin the second round of the DKG process.
+    ///
+    /// ### Parameters
+    /// - `dwallet_cap`: The capability for the associated dWallet.
+    /// - `public_key_share_and_proof`: Public key share and proof from the first round.
+    /// - `first_round_output`: Output from the first DKG round.
+    /// - `first_round_session_id`: Session ID of the first DKG round.
+    public fun launch_dkg_second_round(
+        dwallet_cap: &DWalletCap,
+        public_key_share_and_proof: vector<u8>,
+        first_round_output: vector<u8>,
```

Comment:
We should never trust the user to submit decentralized party outputs. This vulnerability is reported in #436 and fixed.